### PR TITLE
[WIP][SPARK-36180][SQL] Support TimestampNTZ type in Hive metastore

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Support TimestampNTZ type in Hive metastore(HMS) by mapping the TimestampNTZ type to Timestamp in hive. To distinguish TimestampType and TimestampNTZType when reading HMS, Spark SQL will choose the schema version written by Spark SQL when the catalog strings from HMS and Spark SQL are consistent.

### Why are the changes needed?

To support TimestampNTZ type in Hive metastore
### Does this PR introduce _any_ user-facing change?

No, TimestampNTZ is not released yet
### How was this patch tested?

UT

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @beliefer as well

**Review by yaooqinn — DISMISSED** [review]

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 1]
```diff
@@ -130,6 +131,24 @@ abstract class SQLQuerySuiteBase extends QueryTest with SQLTestUtils with TestHi
     }
   }
 
+  test("SPARK-36180: Support TimestampNTZ type in Hive") {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:134]: does this test fail before this PR? I thought we can already save tables with NTZ column into HMS, in hive-incompatible way (Hive can't read this table)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/TimestampNTZType.scala [hunk 2]
```diff
@@ -52,6 +52,8 @@ class TimestampNTZType private() extends DatetimeType {
 
   override def typeName: String = "timestamp_ntz"
 
+  override def catalogString: String = "timestamp"
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/TimestampNTZType.scala:55]: I need more thinking on this. Marking this one as WIP

**Comment by beliefer** [comment]
I don't know the reason. It seems the feature support restart.
I don't know the distinct between this PR and https://github.com/apache/spark/pull/39827.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:134]: Yes it fails with the following erorr
```
org.apache.hadoop.hive.ql.metadata.HiveException: java.lang.IllegalArgumentException: Error: type expected at the position 10 of 'timestamp:timestamp_ntz' but 'timestamp_ntz' is found.
org.apache.spark.sql.AnalysisException: org.apache.hadoop.hive.ql.metadata.HiveException: java.lang.IllegalArgumentException: Error: type expected at the position 10 of 'timestamp:timestamp_ntz' but 'timestamp_ntz' is found.
	at org.apache.spark.sql.hive.HiveExternalCatalog.withClient(HiveExternalCatalog.scala:111)
	at org.apache.spark.sql.hive.HiveExternalCatalog.createTable(HiveExternalCatalog.scala:245)
	at org.apache.spark.sql.catalyst.catalog.ExternalCatalogWithListener.createTable(ExternalCatalogWithListener.scala:94)
	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.createTable(SessionCatalog.scala:402)
	at org.apache.spark.sql.hive.execution.CreateHiveTableAsSelectCommand.run(CreateHiveTableAsSelectCommand.scala:77)
```

**Comment by gengliangwang** [comment]
> I don't know the reason. It seems the feature support restart.

We are going to release TimestampNTZ in Spark 3.4.0.

> Could you tell me the distinct between this PR and https://github.com/apache/spark/pull/39827 ?

Do you mean the difference from https://github.com/apache/spark/pull/34647? This one is trying to change the catalogString of TimestampNTZ and compare catalog strings of Hive table schema and Spark written schema. From the test results, I think this is a bad approach.
Note that we don't have to support this for now. I think supporting all the built-in file sources (except text) is enough for the 3.4.0 release.

**Pull request closed without merging**